relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -0,0 +1,441 @@
+    /// A set of Glob patterns for which the rule applies, empty applies to all
+    #[serde(default)]
+    pub releases: GlobPatterns,
+    /// A set of user_segments for which the rule applies, empty applies to all
+    #[serde(default)]
+    pub user_segments: Vec<String>,
+    /// the sampling rate for trace matching this rule
+    pub sample_rate: f64,
+}
+
+/// Represents the dynamic sampling configuration available to a project.
+/// Note: This comes from the organization data
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SamplingConfig {
+    /// The sampling rules for the project
+    pub rules: Vec<SamplingRule>,
+}
+
+/// TraceContext created by the first Sentry SDK in the call chain
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct TraceContext {
+    /// IID created by SDK to represent the current call flow
+    /// TODO RaduW. Should we deserialize directly into an Uuid ?
+    /// Also since trace_id and public_key are compulsory for tracing functionality should
+    /// they be an option or not ?
+    pub trace_id: Uuid,
+    /// The project key
+    pub public_key: ProjectKey,
+    /// the release
+    #[serde(default)]
+    pub release: Option<String>,
+    /// the user segment
+    #[serde(default)]
+    pub user_segment: Option<String>,
+}
+
+impl TraceContext {
+    /// Returns the decision of whether to sample or not a trace based on the configuration rules
+    /// If None then a decision can't be made either because of an invalid of missing trace context or
+    /// because no applicable sampling rule could be found.
+    pub fn should_sample(&self, config: &SamplingConfig, project_id: ProjectId) -> Option<bool> {
+        let rule = get_matching_rule(config, self, project_id)?;
+        let rate = pseudo_random_from_trace_id(self.trace_id)?;
+        Some(rate < rule.sample_rate)
+    }
+}
+
+/// Tests whether a rule matches a trace context
+fn matches(rule: &SamplingRule, context: &TraceContext, project_id: ProjectId) -> bool {
+    // match against the project
+    if !rule.project_ids.is_empty() && !rule.project_ids.contains(&project_id) {
+        return false;
+    }
+    // match against the release
+    if !rule.releases.is_empty() {
+        match context.release {
+            None => return false,
+            Some(ref release) => {
+                if !rule.releases.is_match(release) {
+                    return false;
+                }
+            }
+        }
+    }
+    // match against the user_segment
+    if !rule.user_segments.is_empty() {
+        match context.user_segment {
+            None => return false,
+            Some(ref user_segment) => {
+                if !rule.user_segments.contains(user_segment) {
+                    return false;
+                }
+            }
+        }
+    }
+    true
+}
+
+fn get_matching_rule<'a>(
+    config: &'a SamplingConfig,
+    context: &TraceContext,
+    project_id: ProjectId,
```

Comment:
Often, we're passing around a `TraceContext` and a `ProjectId`. We had something similar for quotas. 

What do you think about introducing a new type that combines both of them transparently? Something like `ResolvedTraceContext`?

Reply:
It would be nice, if we could use it all over. Unfortunately I get TraceContext from the envelope and the project Id from the project state. I could create the suggested Struct and use it as an internal struct inside dynamic_sampling.rs but it will only be used by `matches` and `get_matching_rule` and possibly (if I remove the function from TraceContext by) `TraceContext::should_sample`. 

To me it doesn't seem worth while introducing another structure, but I can do it if you prefer.